### PR TITLE
Add facility to the PJRT plugin to save artifacts.

### PR DESCRIPTION
**Pull request by stellaraccident**
This can be enabled in the default build via IREE_PJRT_SAVE_ARTIFACTS environment variable. Some effort has been made to allow this to be extended for custom platforms and integration of reproducers.

### CODE REVIEW

File: pjrt-plugin/iree/integrations/pjrt/common/dylib_platform.cc [hunk 1]
```diff
@@ -76,6 +76,22 @@ iree_status_t DylibPlatform::SubclassInitialize() {
   // And initialize the compiler.
   compiler_ = std::make_unique<InprocessCompiler>();
 
+  // Initialize the artifact dumper.
+  auto artifact_path_spec = config_vars().Lookup("SAVE_ARTIFACTS");
+  if (!artifact_path_spec) {
```
Comment (jpienaar) [inline pjrt-plugin/iree/integrations/pjrt/common/dylib_platform.cc:81]: Prefer positive logic

File: pjrt-plugin/iree/integrations/pjrt/common/platform.cc [hunk 2]
```diff
@@ -50,6 +53,134 @@ void Logger::error(std::string_view message) {
...
+        message.append(p);
+        logger_.debug(message);
+      }
+    }
+
+    written_paths_.clear();
+  }
+
+ private:
+  Logger& logger_;
```
Comment (jpienaar) [inline pjrt-plugin/iree/integrations/pjrt/common/platform.cc:148]: I think it would be smaller class layout if reorganized, but that's not critical.

**Review by jpienaar — COMMENTED** [review]
Nice, thanks!

File: pjrt-plugin/iree/integrations/pjrt/common/platform.cc (hunk 2 above)
Comment (stellaraccident) [reply pjrt-plugin/iree/integrations/pjrt/common/platform.cc:148]: I can usually spot those but missing this time. Looks like a bunch of pointer sized things, followed by bool... which is the right way, right?

**Pull request merged**